datafusion/physical-expr/src/expressions/case.rs

```diff
@@ -344,7 +344,16 @@ impl CaseExpr {
     fn case_column_or_null(&self, batch: &RecordBatch) -> Result<ColumnarValue> {
         let when_expr = &self.when_then_expr[0].0;
         let then_expr = &self.when_then_expr[0].1;
-        if let ColumnarValue::Array(bit_mask) = when_expr.evaluate(batch)? {
+
+        let when_expr_value = when_expr.evaluate(batch)?;
+        let when_expr_value = match when_expr_value {
```

Comment:
this is correct, but it effectively expands out the array even when the argument is a constant. We can probably make it more efficient by handling the other case too

Reply:
As a follow up, I have added one in
- https://github.com/apache/datafusion/pull/14159